src/EvenVCO.cpp

```diff
@@ -49,102 +49,184 @@ struct EvenVCO : Module {
+				doubleSawMinBlep[c].insertDiscontinuity(crossing, -2.f);
+				squareMinBlep[c].insertDiscontinuity(crossing, -2.f);
+				sawMinBlep[c].insertDiscontinuity(crossing, -2.f);
+				halfPhase[c] = false;
+			}
+		}
+
+		float_4 triSquareMinBlepOut[4] = {};
+		float_4 doubleSawMinBlepOut[4] = {};
+		float_4 sawMinBlepOut[4] = {};
+		float_4 squareMinBlepOut[4] = {};
+
+		float_4 triSquare[4] = {};
+		float_4 sine[4] = {};
+		float_4 doubleSaw[4] = {};
+
+		float_4 even[4] = {};
+		float_4 saw[4] = {};
+		float_4 square[4] = {};
+		float_4 triOut[4] = {};
+
+		for (int c = 0; c < channels; c++) {
+			triSquareMinBlepOut[c / 4].s[c % 4] = triSquareMinBlep[c].process();
+			doubleSawMinBlepOut[c / 4].s[c % 4] = doubleSawMinBlep[c].process();
+			sawMinBlepOut[c / 4].s[c % 4] = sawMinBlep[c].process();
+			squareMinBlepOut[c / 4].s[c % 4] = squareMinBlep[c].process();
 		}
 
 		// Outputs
-		float triSquare = (phase < 0.5) ? -1.f : 1.f;
-		triSquare += triSquareMinBlep.process();
-
-		// Integrate square for triangle
-		tri += 4.f * triSquare * freq * args.sampleTime;
-		tri *= (1.f - 40.f * args.sampleTime);
-
-		float sine = -std::cos(2*M_PI * phase);
-		float doubleSaw = (phase < 0.5) ? (-1.f + 4.f*phase) : (-1.f + 4.f*(phase - 0.5));
-		doubleSaw += doubleSawMinBlep.process();
-		float even = 0.55 * (doubleSaw + 1.27 * sine);
-		float saw = -1.f + 2.f*phase;
-		saw += sawMinBlep.process();
-		float square = (phase < pw) ? -1.f : 1.f;
-		square += squareMinBlep.process();
-
-		// Set outputs
-		outputs[TRI_OUTPUT].setVoltage(5.f*tri);
-		outputs[SINE_OUTPUT].setVoltage(5.f*sine);
-		outputs[EVEN_OUTPUT].setVoltage(5.f*even);
-		outputs[SAW_OUTPUT].setVoltage(5.f*saw);
-		outputs[SQUARE_OUTPUT].setVoltage(5.f*square);
+		outputs[TRI_OUTPUT].setChannels(channels);
+		outputs[SINE_OUTPUT].setChannels(channels);
+		outputs[EVEN_OUTPUT].setChannels(channels);
+		outputs[SAW_OUTPUT].setChannels(channels);
+		outputs[SQUARE_OUTPUT].setChannels(channels);
+
+		for (int c = 0; c < channels; c += 4) {
+
+			triSquare[c / 4] = simd::ifelse((phase[c / 4] < 0.5f), -1.f, +1.f);
+			triSquare[c / 4] += triSquareMinBlepOut[c / 4];
+
+			// Integrate square for triangle
+
+			tri[c / 4] += (4.f * triSquare[c / 4]) * (freq[c / 4] * args.sampleTime);
+			tri[c / 4] *= (1.f - 40.f * args.sampleTime);
+			triOut[c / 4] = 5.f * tri[c / 4];
+
+			sine[c / 4] = 5.f * simd::cos(2 * M_PI * phase[c / 4]);
+
+			doubleSaw[c / 4] = simd::ifelse((phase[c / 4] < 0.5), (-1.f + 4.f * phase[c / 4]), (-1.f + 4.f * (phase[c / 4] - 0.5f)));
+			doubleSaw[c / 4] += doubleSawMinBlepOut[c / 4];
+			doubleSaw[c / 4] *= 5.f;
+
+			even[c / 4] = 0.55 * (doubleSaw[c / 4] + 1.27 * sine[c / 4]);
+			saw[c / 4] = -1.f + 2.f * phase[c / 4];
+			saw[c / 4] += sawMinBlepOut[c / 4];
+			saw[c / 4] *= 5.f;
+
+			square[c / 4] = simd::ifelse((phase[c / 4] < pw[c / 4]),  -1.f, +1.f);
+			square[c / 4] += squareMinBlepOut[c / 4];
+			square[c / 4] *= 5.f;
+
```

Comment:
won't this pulse wave have a lot of DC if it just swings from -5 to +5?

Reply:
[ I know a lot of VCOs put out a ton of DC on their pulse outputs (and a little bit on their saw). But it does make for a lot of clicks and pops if there is DC. These ones don't have any DC: https://github.com/squinkylabs/Demo

Reply:
Yeah absolutely, I added an optional (default disabled) DC block to Chopping Kinky for that exact reason (the hardware version is DC coupled, so wanted to be as close as possible, but still). As for EvenVCO, I'd actually like to overhaul it properly at some stage (including this, adding H. sync, and maybe some more performance improvements). However I'm keen to not delay this release too much, so maybe could do that as v1.1.1 for example?

Your EV3 is/will be a super useful reference for that process.

Reply:
ah, once again, very nice. yeah, plans sound good. btw - don't look at my sync implementation in EV3 for ideas. the one in Fundamental VCO 1 would be a much better place to start. On a super minor point, I get rid of DC on the pulse (and saw) by estimating the expected DC and subtracting. Cheaper and better than using a filter, imao. But three cheers for filter! One less time I will have to hear those distinctive thumps that plagues VCV videos.